src/tqec/compile/substitute.py

```diff
@@ -0,0 +1,203 @@
+from tqec.position import Direction3D
+from tqec.sketchup.block_graph import PipeType
+
+
+@dataclass(frozen=True)
+class SubstitutionKey:
+    """The key for selecting the substitution rule.
+
+    Substitution rules will be selected based on the key to apply the substitution
+    during the block graph compilation.
+
+    Attributes:
+        spec1: the cube specification of the first cube. By convention, the cube
+            corresponding to `spec1` should have a smaller position than the cube
+            corresponding to `spec2`.
+        spec2: the cube specification of the second cube.
+        pipe_type: the type of the pipe connecting the two cubes.
+    """
+
+    spec1: CubeSpec
+    spec2: CubeSpec
+    pipe_type: PipeType
+
+
+class SubstitutionRule(Protocol):
+    def __call__(
+        self,
+        key: SubstitutionKey,
+        block1: CompiledBlock,
+        block2: CompiledBlock,
+    ) -> tuple[CompiledBlock, CompiledBlock]:
+        """Substitute the two blocks with the given key.
+
+        Args:
+            key: `SubstitutionKey` that includes the necessary information for the
+                substitution, including the cube specifications and the pipe type that
+                is connecting the two blocks.
+            block1: the first block to substitute.
+            block2: the second block to substitute.
+        """
+        ...
+
+
+def default_substitution_rule(
+    key: SubstitutionKey,
+    block1: CompiledBlock,
+    block2: CompiledBlock,
+) -> tuple[CompiledBlock, CompiledBlock]:
+    """The default substitution rule for the block graph compilation."""
+    # 1. The spatial junctions are not supported yet
+    if key.spec1.is_spatial_junction or key.spec2.is_spatial_junction:
+        raise TQECException("Spatial junctions are not supported yet.")
+    # 2. Substitute the time-direction pipe
+    if key.pipe_type.direction == Direction3D.Z:
+        return _substitute_in_time_direction(key, block1, block2)
+    # 3. Substitute the space-direction pipe
+    else:
+        return _substitute_in_space_with_usual_cubes(key, block1, block2)
+
+
+DEFAULT_SUBSTITUTION_RULES: defaultdict[SubstitutionKey, SubstitutionRule] = (
+    defaultdict(lambda: default_substitution_rule)
+)
+
+
+def _substitute_in_time_direction(
+    key: SubstitutionKey,
+    block1: CompiledBlock,
+    block2: CompiledBlock,
+) -> tuple[CompiledBlock, CompiledBlock]:
+    assert key.pipe_type.direction == Direction3D.Z, "Pipe direction must be Z."
+    # Substitute the final layer of the bottom block with the bulk layer
+    bottom_block = block1.with_updated_layer(
+        plaquettes_to_update=deepcopy(block1.layers[1].collection),
+        layer_to_update=block1.num_layers - 1,
+    )
+    # Substitute the first layer of the top block with the bulk layer
+    top_block = block2.with_updated_layer(
+        plaquettes_to_update=deepcopy(block2.layers[1].collection),
+        layer_to_update=0,
+    )
```

Comment:
We need it here because later on we will add reset/measurement on data qubits at certain layers. If we do not deep copy it, we will end up adding unwanted operations to some layers. I tested after removing it and can be sure about this.

Reply:
This is necessary for substitution of "L" shape blocks to work.

Reply:
If Plaquette is immutable, a copy should be enough no?

Reply:
But it is not immutable due to the need of adding measurements and resets. Let's leave it like this for the moment, this is probably something we will come back to later